File: qForDevelopersPython/q-demo-03/log-cleaner/src/logcleaner/parsing.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Optional


@dataclass(frozen=True)
class LogEvent:
    timestamp: str
    level: str
    component: str
    message: str
    source: str


VALID_LEVELS = {"DEBUG", "INFO", "WARN", "ERROR"}


def parse_line(source: str, line: str) -> Optional[LogEvent]:
    """
    Parse one log line.

    Expected format:
      <timestamp> <level> <component> <message...>

    Returns None for lines that don't match expectations.
    """
    parts = line.split(" ", 3)
    if len(parts) < 4:
        return None

    timestamp, level, component, message = parts[0], parts[1], parts[2], parts[3]
    level = level.upper()
    if level not in VALID_LEVELS:
        return None

    component = component.strip()
    message = message.strip()

    return LogEvent(
        timestamp=timestamp,
        level=level,
        component=component,
        message=message,
        source=source,
    )

File: qForDevelopersJava/demo-03/src/main/java/logcleaner/Parsing.java
package logcleaner;

import logcleaner.model.LogEvent;

import java.util.Optional;

public final class Parsing {
    private Parsing() {}

    /**
     * Parse a log line like:
     * 2026-02-15T10:12:30Z INFO auth User logged in
     *
     * Returns Optional.empty() if the line is malformed.
     */
    public static Optional<LogEvent> parseLine(String sourceFile, String line) {
        if (line == null || line.isBlank()) return Optional.empty();

        // Split into at most 3 parts: timestamp, level, and "rest"
        String[] parts = line.trim().split("\\s+", 3);
        if (parts.length < 3) return Optional.empty();

        String timestamp = parts[0];
        String level = parts[1];
        String rest = parts[2];

        String component = rest;     
        String message = "";         

        return Optional.of(new LogEvent(sourceFile, timestamp, level, component, message));
    }
}